src/Mvc/Mvc.Formatters.Xml/src/XmlDataContractSerializerOutputFormatter.cs

```diff
@@ -254,25 +259,38 @@ public override async Task WriteResponseBodyAsync(OutputFormatterWriteContext co
 
             var dataContractSerializer = GetCachedSerializer(wrappingType);
 
-            // Opt into sync IO support until we can work out an alternative https://github.com/aspnet/AspNetCore/issues/6397
-            var syncIOFeature = context.HttpContext.Features.Get<Http.Features.IHttpBodyControlFeature>();
-            if (syncIOFeature != null)
+            var responseStream = GetResponseStream(context.HttpContext);
+
+            try
             {
-                syncIOFeature.AllowSynchronousIO = true;
+                await using (var textWriter = context.WriterFactory(responseStream, writerSettings.Encoding))
+                {
+                    using (var xmlWriter = CreateXmlWriter(context, textWriter, writerSettings))
+                    {
+                        dataContractSerializer.WriteObject(xmlWriter, value);
+                    }
+                }
             }
-
-            using (var textWriter = context.WriterFactory(context.HttpContext.Response.Body, writerSettings.Encoding))
+            finally
             {
-                using (var xmlWriter = CreateXmlWriter(context, textWriter, writerSettings))
+                if (responseStream is FileBufferingWriteStream fileBufferingWriteStream)
                 {
-                    dataContractSerializer.WriteObject(xmlWriter, value);
+                    await fileBufferingWriteStream.DisposeAsync();
                 }
```

Comment:
yeah. That's what I mean.

Reply:
You want to separate flush and dispose, you always want to dispose but you don't want to flush under error conditions. You also don't have to sniff the stream type in order to call FlushAsync.

Reply:
If we require the caller to force calling a `FlushAsync`, I think we need to rethink the design of having this be pass-thru. It would make much more sense to just have this be an entirely buffering stream and have the caller `CopyToAsync`

Reply:
Razor does it's own buffering. It doesn't require this.